### DISCUSSION TITLE
Expose is_required in extension API for virtual methods

### DISCUSSION THREAD

**Discussion by TitanNano** [godotengine/godot-proposals#9982] [Engine Core]
The engine currently does not expose if a virtual method is required to be implemented or not. The only indicator for a method to be required is if there is any call to it via `GDVIRTUAL_REQUIRED_CALL` at runtime. 

It would be beneficial to GDExtensions if virtual methods were marked as required and only then `GDVIRTUAL_REQUIRED_CALL` could be called on them. The engine should then also expose a `is_required` field in `extension_api.json`. 

Bindings like [godot-rust/gdext](https://github.com/godot-rust/gdext/) could then handle these required methods appropriately, e.g. fail the compilation if a method has not been implemented. 

EDIT: Related issue in https://github.com/godot-rust/gdext/issues/192#issuecomment-2170971885

**Comment by dsnopek** [comment]
This is an interesting idea!

I guess we could include whether or not it's required in the definition (maybe via an extension to the `GDVIRTUAL*()` macros?), and then get rid of `GDVIRTUAL_REQUIRED_CALL()` and just have `GDVIRTUAL_CALL()` do the right thing depending on the definition. This could use `constexpr` to avoid adding any extra overhead. So, I don't think there's any technical problems with doing it.

Then I guess it'd be a question of whether or not we think it's worth changing all the `GDVIRTUAL*()` declarations in the code base in order to add this extra information.

@Bromeon What do you think about this?

**Reply by Bromeon** [reply]
I think an `is_required` flag is a great idea. Some APIs like [`ScriptExtension`](https://docs.godotengine.org/en/stable/classes/class_scriptextension.html) have a ton of virtual methods, and checking for runtime errors one-by-one can be quite tedious.

Regarding implementation, I'm not very familiar with the `GDVIRTUAL_REQUIRED_CALL` and `GDVIRTUAL*` macros, so it's hard to judge how much effort such a change would be. But if you believe it's reasonably possible, I'd say we could give it a try 🙂 or do you see any downsides?

**Reply by AThousandShips** [reply]
I don't think it's feasible to detect it via the calls, and that if we want it to be automated we should change the declaration macros, but that adds quite a bit of extra complexity

I.e. like `GDVIRTUAL` macros with a required part along with the argument and return

However there *could* be times when a method is required but wouldn't always be called with it, so I'd keep the call part separate

And we *could* have a check so that a bind must be marked required if called that way

I'll look at the code when I have some time and energy for it and see if I can come up with some ideas

**Reply by dsnopek** [reply]
For the implementation, I was thinking we could change the declaration macros, so we'd have like `GDVIRTUAL*_REQUIRED()` (ex `GDVIRTUAL1RC_REQUIRED(int, int)` or whatever), and add something like `static constexpr bool _gdvirtual_##m_name##_required = true; // or false, for the non-required one`.

This value could be used by the `GDVIRTUAL_BIND()` macro to add the required information to `ClassDB`. And, we could either have `GDVIRTUAL_CALL()` look at that or just generate the call method from the declaration macro to either be required or not rather than using a template argument.

Anyway, I think the implementation should be fairly easy.

> But if you believe it's reasonably possible, I'd say we could give it a try 🙂 or do you see any downsides?

The biggest challenge that I personally see is just the work of changing all the `GDVIRTUAL*()` and `GDVIRTUAL_CALL()` usages throughout the code base, and not making any mistakes. :-)

> However there _could_ be times when a method is required but wouldn't always be called with it, so I'd keep the call part separate

Perhaps. If there isn't a case like that already, I'm not sure we should go out of our way to account for it, though.

**Reply by TitanNano** [reply]
> The biggest challenge that I personally see is just the work of changing all the GDVIRTUAL*() and GDVIRTUAL_CALL() usages throughout the code base, and not making any mistakes. :-)

Does it make sense to keep `GDVIRTUAL_REQUIRED_CALL` and make it only compatible with `GDVIRTUAL*_REQUIRED()` so that the right pair has to always be used? I don't know if that is possible.

**Reply by dsnopek** [reply]
> Does it make sense to keep `GDVIRTUAL_REQUIRED_CALL` and make it only compatible with `GDVIRTUAL*_REQUIRED()` so that the right pair has to always be used?

Personally, I don't see the sense in keeping it if the right pair always has to be used.

**Reply by dsnopek** [reply]
I did a quick prototype implementation in draft PR https://github.com/godotengine/godot/pull/93311

I didn't end up adding the `static constexpr bool _gdvirtual_##m_name##_required` variable, and instead added a new `METHOD_FLAG_VIRTUAL_REQUIRED`, since we're storing the virtual methods in `ClassDB` as `MethodInfo`.

**Comment by TitanNano** [comment]
This has been implemented in https://github.com/godotengine/godot/pull/93311
